src/librustc_typeck/check/demand.rs

```diff
@@ -425,16 +425,29 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                                 }
                             }
                         }
+
+                        let mut sugg = sugg_expr.as_str();
+                        if let hir::ExprKind::MethodCall(_segment, _sp, _args) = &expr.node {
+                            let clone_path = "std::clone::Clone::clone";
+                            if let Some(true) = self.tables.borrow()
+                                .type_dependent_def_id(expr.hir_id)
+                                .map(|did| self.tcx.def_path_str(did).as_str() == clone_path)
+                            {
+                                // If this expression had a clone call when suggesting borrowing
+                                // we want to suggest removing it because it'd now be unecessary.
+                                sugg = sugg_expr.trim_end_matches(".clone()");
```

Comment:
Can't you check that the `MethodCall` has one argument and use that argument's `Span` instead of `sp`, in `if let Ok(src) = cm.span_to_snippet(sp) {`?
I'd find that much cleaner than doing *anything* with strings.